app/vmselect/promql/rollup_result_cache.go

```diff
@@ -607,7 +608,7 @@ func mergeTimeseries(qt *querytracer.Tracer, a, b []*timeseries, bStart int64, e
 		tmp.denyReuse = true
 		tmp.Timestamps = sharedTimestamps
 		tmp.Values = make([]float64, 0, len(tmp.Timestamps))
-		tmp.MetricName.MoveFrom(&tsB.MetricName)
+		tmp.MetricName.CopyFrom(&tsB.MetricName)
```

Comment:
This function has the following comment:
```
// Postconditions:
// - a and b cannot be used after returning from the call.
```
But we do use `b` after this function call here https://github.com/VictoriaMetrics/VictoriaMetrics/blob/d389a4fcf34528bb8ff9086813f20667de1742d8/app/vmselect/promql/eval.go#L1626-L1631

Should we not use `b` instead?

Reply:
It's not safe also. PutSeries modifies timeseries, it aligns timestamps end to the now()-step-cacheOffSet. 

 Probably, there is better solution.

Reply:
> But we do use b after this function call here

It looks like this is the real cause of the bug - the `rvs` must be stored in the case instead of `tss`. Also the check for `tss != nil` is unneeded here, since `PutSeries()` contains a check for empty series.

But even after storing the proper series to the cache, there is a valid (yet rare) possibility when duplicate series may be returned from query results at then passed to `mergeTimeseries()` in `tss` arg at the line 1626. For example, when `quantile(0.6, rate({__name__=~"http_requests_total|http_request_errors_total"}[5m]))` query is executed, then `tss` will contain results for `rate({__name__=~"http_requests_total|http_request_errors_total"}[5m])`. These results do not have metric name (it is stripped by `rate()`), while they may have identical sets of labels. In this case the `mergeTimeseries()` cannot determine how to merge cached series with these duplicate series. This case should be gracefully handled instead of returning error to the user. For example, in this case the query must be executed with disabled cache.

I implemented both fixes in the commit 2cbdb1db22cff7286cf056504a1179536c322f6e .

Reply:
@valyala  PutSeries to cache applies removes timestamps and values outside deadline:

```
// Remove values up to currentTime - step - cacheTimestampOffset,
	// since these values may be added later.
	timestamps := tss[0].Timestamps
	deadline := (time.Now().UnixNano() / 1e6) - ec.Step - cacheTimestampOffset.Milliseconds()
	i := len(timestamps) - 1
	for i >= 0 && timestamps[i] > deadline {
		i--
	}
```

 Query will return nothing for the latest 5 min  + expr step

Reply:
Yeah, the code inside [`rollupResultCache.PutSeries()`](https://github.com/VictoriaMetrics/VictoriaMetrics/blob/c8e6e47e2a989b0f2aabbc179436d37b7bbd755c/app/vmselect/promql/rollup_result_cache.go#L350-L443) is tricky. In reality, this code makes a shallow copy `tss` series if it needs to drop samples outside `cacheTimestampOffset`:

https://github.com/VictoriaMetrics/VictoriaMetrics/blob/c8e6e47e2a989b0f2aabbc179436d37b7bbd755c/app/vmselect/promql/rollup_result_cache.go#L395-L400

This means that the original `tss` series remain unmodified.